Add docs on using ExprP() for custom WHERE sql statement (#2520)
ent/ent: doc/md/predicates.md

Custom predicates can be useful if you want to write your own dialect-specific logic or to control the executed queries.

For example, in order to use built-in SQL functions such as `DATE()`, use one of the following options:

1. Pass a dialect-aware predicate function using the `sql.P` option:

```go
users := client.User.Query().
	Select("id").

> 	Where(P(func(b *Builder) {
> 		b.WriteString("DATE(").Ident("last_login_at").WriteByte(')').WriteOp(OpGTE).Arg(value)
> 	}))

Reviewer: ```suggestion
	Where(sql.P(func(b *sql.Builder) {
		b.WriteString("DATE(").Ident("last_login_at").WriteByte(')').WriteOp(OpGTE).Arg(value)
	})).
	AllX(ctx)
```

Reviewer: Also, you can't use `sql.P` inside `UserQuery.Where`. I can edit this example later and land it if you prefer. Thanks

Custom predicates can be useful if you want to write your own dialect-specific logic or to control the executed queries.

For example, in order to use built-in SQL functions such as `DATE()`, use one of the following options:

1. Pass a dialect-aware predicate function using the `sql.P` option:

```go
users := client.User.Query().

> 	Select("id").

Reviewer: ```suggestion
	Select(user.FieldID).
```

	Where(sql.P(func(b *sql.Builder) {
		b.WriteString("DATE(").Ident("last_login_at").WriteByte(')').WriteOp(OpGTE).Arg(value)
	})).
	AllX(ctx)
```

The above code will produce the following SQL query:

```sql
SELECT `id` FROM `users` WHERE DATE(`last_login_at`) >= ?
```

2. Inline a predicate expression using the `ExprP()` option:

```go
users := client.User.Query().
	Select("id").
	Where(func(s *sql.Selector) {
		s.Where(sql.ExprP("DATE(last_login_at >= ?", value))
	}).
	AllX(ctx)
```

> 
> But note that `last_login_at` is now hardcoded, unlike the option number 1. 
> 
> The above code will produce the same SQL query:

Reviewer: ```suggestion

The above code will produce the same SQL query:
```

```sql
SELECT `id` FROM `users` WHERE DATE(`last_login_at`) >= ?
```

> 
> You may also perform a more complex sql query, for example `DATE_ADD()`
> 
> ```go
> events := client.Event.Query().
> 	Where(func(s *sql.Selector) {
> 		s.Where(sql.ExprP("DATE_ADD(date, INTERVAL duration MINUTE) BETWEEN ? AND ?", start, end))
> 	}).
> 	AllX(ctx)
> ```
> 
> The above code will produce the following SQL query:
> 
> ```sql
> SELECT `events`.`id`, `events`.`date`, `events`.`duration` 
> FROM events WHERE DATE_ADD(date, INTERVAL duration MINUTE) BETWEEN ? AND ?
> ```
> 

Reviewer: ```suggestion

```

I feel like the 2 examples above are enough, also this PR sits for a long time and I really want to merge it.

Custom predicates can be useful if you want to write your own dialect-specific logic or to control the executed queries.

> For example, if you want to use builtin sql functions such as a `DATE()` function to a `WHERE` statement.
> It can be achieved by using either `P()` or `ExprP()`.

Reviewer: ```suggestion
For example, in order to use built-in SQL functions such as `DATE()`, use one of the following options:
```

> 1. `P()`
> 
> Using `P()` gives more flexibility in how you arrange the strings.

Reviewer: ```suggestion
1\. Pass a dialect-aware predicate function using the `sql.P` option:
```

```go
Select("id").
From(Table("users")).

> Where(P(func(b *Builder) {

Reviewer: Align this example with the one below.

>     b.WriteString("DATE(").Ident("last_login_at").WriteString(") >= ").Arg("2022-05-03")

Reviewer: ```suggestion
    b.WriteString("DATE(").Ident("last_login_at").WriteByte(')').WriteOp(OpGTE).Arg(value)
```

}))
```

The above code will produce the following SQL query:

```sql
SELECT `id` FROM `users` WHERE DATE(`last_login_at`) >= ?
```

> 2. `ExprP()`
> 
> The same thing can be achieved with

Reviewer: ```suggestion
2\. Invalid a predicate expression using the `ExprP()` option:
```

Author: Hi @a8m, thank you for all the feedback. I have made the changes as you suggested.

However for this, I do not understand the wording ` Invalid a predicate expression...`. Are we invalidating anything?

Reviewer: Sorry for the confusion. I think I meant “inline”, but accidentally wrote “invalid”.

Author: no worries! pushed all commits for review

```go
users := client.User.Query().

>     Where(func(s *ent.Selector) {

Reviewer: ```suggestion
    Where(func(s *sql.Selector) {
```

>         s.Where(sql.ExprP("DATE(last_login_at >= ?", dateVar))		

Reviewer: ```suggestion
        s.Where(sql.ExprP("DATE(last_login_at >= ?", value))		
```

    }).
    AllX(ctx)
```

The above code will produce the following SQL query:

```sql
SELECT `users`.`id`, `users`.`last_login_at` FROM users WHERE DATE(last_login_at) >= ?;
```

You may also perform a more complex sql query, for example `DATE_ADD()`

```go

> events := r.ent.Event.Query().

Reviewer: ```suggestion
events := client.Event.Query().
```

>     Where(func(s *entsql.Selector) {

Reviewer: ```suggestion
    Where(func(s *sql.Selector) {
```

>         s.Where(entsql.ExprP("DATE_ADD(date, INTERVAL duration MINUTE) BETWEEN ? AND ?", startDateVar, endDateVar))

Reviewer: ```suggestion
        s.Where(entsql.ExprP("DATE_ADD(date, INTERVAL duration MINUTE) BETWEEN ? AND ?", start, end))
```

SELECT DISTINCT `pets`.`id`, `pets`.`owner_id`, `pets`.`name`, `pets`.`age`, `pets`.`species` FROM `pets` WHERE `name` LIKE '_B%'
```

> #### Escape hatch for `WHERE`

Reviewer: I'd merge it with the custom predicate section above. Also, the `ExprP` is one way of doing it, you can also achieve the same thing with `P(func(b *sql.Builder) { ... })` - which gives more control over how identifiers and placeholders are formatted. I can suggest another example option for writing it if you like.

Author: Like this?

```go
{
	input: Select("id").From(Table("users")).Where(P(func(b *Builder) {
		b.WriteString("DATE(last_login_at) >= ?").Arg("2022-05-03")
	})),
	wantQuery: "SELECT `id` FROM `users` WHERE Date(last_login_at) >= ?",
	wantArgs:  []interface{}{"2022-05-03"},
},
```

Author: I think it should be this instead

```go
{
	input: Select("id").
		From(Table("users")).
		Where(P(func(b *Builder) {
			b.WriteString("DATE(").Ident("last_login_at").WriteString(") >= ").Arg("2022-05-03")
		})),
	wantQuery: "SELECT `id` FROM `users` WHERE DATE(`last_login_at`) >= ?",
	wantArgs:  []interface{}{"2022-05-03"},
},
```
